src/Features/Core/Portable/Completion/CompletionServiceWithProviders.cs

```diff
@@ -0,0 +1,402 @@
+using Roslyn.Utilities;
+
+
+namespace Microsoft.CodeAnalysis.Completion
+{
+    /// <summary>
+    /// A subtype of <see cref="CompletionService"/> that aggregates completions from one or more <see cref="CompletionProvider"/>s.
+    /// </summary>
+    public abstract class CompletionServiceWithProviders : CompletionService
+    {
+        private static readonly Func<string, List<CompletionItem>> s_createList = _ => new List<CompletionItem>();
+        private IEnumerable<Lazy<CompletionProvider, CompletionProviderMetadata>> _importedProviders;
+        private readonly Workspace _workspace;
+
+        protected CompletionServiceWithProviders(Workspace workspace)
+        {
+            _workspace = workspace;
+        }
+
+        public override CompletionRules GetRules()
+        {
+            return CompletionRules.Default;
+        }
+
+        /// <summary>
+        /// Returns the providers always available to the service.
+        /// This does not included providers imported via MEF composition.
+        /// </summary>
+        protected virtual ImmutableArray<CompletionProvider> GetBuiltInProviders()
+        {
+            return ImmutableArray<CompletionProvider>.Empty;
+        }
+
+        private IEnumerable<Lazy<CompletionProvider, CompletionProviderMetadata>> GetImportedProviders()
+        {
+            if (_importedProviders == null)
+            {
+                var language = this.Language;
+                var mefExporter = (IMefHostExportProvider)_workspace.Services.HostServices;
+
+                var providers = ExtensionOrderer.Order(
+                        mefExporter.GetExports<CompletionProvider, CompletionProviderMetadata>()
+                        .Where(lz => lz.Metadata.Language == language)
+                        ).ToList();
+
+                Interlocked.CompareExchange(ref _importedProviders, providers, null);
+            }
+
+            return _importedProviders;
+        }
+
+        private ImmutableArray<CompletionProvider> _testProviders = ImmutableArray<CompletionProvider>.Empty;
+
+        internal void SetTestProviders(IEnumerable<CompletionProvider> testProviders)
+        {
+            _testProviders = testProviders != null ? testProviders.ToImmutableArray() : ImmutableArray<CompletionProvider>.Empty;
+        }
+
+        private class RoleProviders
+        {
+            public ImmutableArray<CompletionProvider> Providers;
+        }
+
+        private readonly ConditionalWeakTable<ImmutableHashSet<string>, RoleProviders> _roleProviders
+            = new ConditionalWeakTable<ImmutableHashSet<string>, RoleProviders>();
+
+        private ImmutableArray<CompletionProvider> GetProvidersByRole(ImmutableHashSet<string> roles)
+        {
+            roles = roles ?? ImmutableHashSet<string>.Empty;
+
+            RoleProviders providers;
+            if (!_roleProviders.TryGetValue(roles, out providers))
+            {
+                providers = _roleProviders.GetValue(roles, _ =>
```

Comment:
I hink there's a GetOrAdd extension (or something like that) so you don't need the "TryGetValue and GetValue" pattern.